src/v/cloud_storage/inventory/inv_consumer.cc

```diff
@@ -0,0 +1,250 @@
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/seastar.hh>
+
+#include <re2/re2.h>
+
+#include <ranges>
+
+namespace ranges = std::ranges;
+namespace views = std::views;
+
+namespace {
+const RE2 path_expr{"^[[:xdigit:]]+/(.*?)/(.*?)/(\\d+)_\\d+/.*?"};
+struct flush_entry {
+    model::ntp ntp;
+    fragmented_vector<uint64_t> hashes;
+};
+
+// Will remove nested, embedded quotes and spaces. The CSV report format we
+// expect should be resilient to this because we only care about path names,
+// which contain neither spaces nor quotes, but this is not a real CSV parser.
+bool drop_chars(char c) { return c != '"' && !std::isspace(c); }
+
+ss::sstring trim_string(ranges::subrange<char*> r) {
+    auto filtered = r | views::filter(drop_chars);
+    return {filtered.begin(), filtered.end()};
+}
+
+} // namespace
+
+namespace cloud_storage::inventory {
+
+inventory_consumer::inventory_consumer(
+  std::filesystem::path hash_store_path,
+  absl::node_hash_set<model::ntp> ntps,
+  size_t max_paths_in_memory)
+  : _hash_store_path{std::move(hash_store_path)}
+  , _ntps{std::move(ntps)}
+  , _max_hash_size_in_memory{max_paths_in_memory} {}
+
+ss::future<> inventory_consumer::consume(
+  ss::input_stream<char> s, gzip_compressed_t is_report_compressed) {
+    std::exception_ptr ep;
+    try {
+        co_await do_consume(std::move(s), is_report_compressed);
+    } catch (...) {
+        ep = std::current_exception();
+    }
+
+    if (ep) {
+        vlog(cst_log.error, "Failed to consume inventory report: {}", ep);
+        co_await stop();
+        std::rethrow_exception(ep);
+    }
+}
+
+ss::future<> inventory_consumer::do_consume(
+  ss::input_stream<char> s, gzip_compressed_t is_report_compressed) {
+    report_parser parser{std::move(s), is_report_compressed};
+    co_await parser.consume([this](const auto& path) {
+        // The row is expected to be in the form:
+        // "bucket","path"
+        auto pieces = parse_row(path);
+        // If a row is malformed we skip it and do not abort processing. This
+        // way the rest of the entries are processed, and if there was a path in
+        // this row it will be recorded as missing.
+        if (pieces.size() != 2) {
+            vlog(cst_log.warn, "unexpected row in inventory report: {}", path);
+            return ss::make_ready_future();
+        }
+        return process_path(pieces[1]);
+    });
+}
+
+std::vector<ss::sstring> inventory_consumer::parse_row(ss::sstring row) {
+    std::vector<ss::sstring> pieces;
+    for (auto p : row | views::split(',') | views::transform(trim_string)) {
+        pieces.emplace_back(p);
+    }
+    return pieces;
+}
+
+ss::future<> inventory_consumer::process_path(ss::sstring path) {
+    if (auto maybe_ntp = ntp_from_path(path);
+        _ntps.contains(maybe_ntp.value())) {
+        const auto& ntp = maybe_ntp.value();
+        auto hash = xxhash_64(path.data(), path.size());
```

Comment:
if we used something like sha256 (or the first half of it) we could probably not even bother with collision detection, right? could simplify some things.

Reply:
I thought since we have a relatively cheap way to handle collisions which should be very few even with 64 bit hash, using `xxhash_64` would use just a quarter of the memory and disk space of sha256.